src/accumulate.jl

```diff
@@ -3,122 +3,138 @@
 ## COV_EXCL_START
 
 # partial scan of individual thread blocks within a grid
-# work-efficient implementation after Blelloch (1990)
+# Parallel Prefix Scan using warp-level primitives
 #
-# number of threads needs to be a power-of-2
+# https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-39-parallel-prefix-sum-scan-cuda
+# 
 #
 # performance TODOs:
-# - shuffle
-# - warp-aggregate atomics
-# - the ND case is quite a bit slower than the 1D case (not using Cartesian indices,
-#   before 35fcbde1f2987023229034370b0c9091e18c4137). optimize or special-case?
-function partial_scan(op::Function, output::CuDeviceArray{T}, input::CuDeviceArray,
-                      Rdim, Rpre, Rpost, Rother, neutral, init,
-                      ::Val{inclusive}=Val(true)) where {T, inclusive}
+# - Multiple elements per thread 
+# - Use warpSize instead of hardcoding and move to GPUArrays
+function partial_scan(op::Function, output::CuDeviceArray{T}, input, aggregates,
+                        Rdim, Rpre, Rpost, Rother, neutral) where {T}
     threads = blockDim().x
     thread = threadIdx().x
-    block = blockIdx().x
+    warps = threads ÷ 32
+    mask = typemax(UInt32)
 
-    temp = @cuDynamicSharedMem(T, (2*threads,))
+    temp = @cuDynamicSharedMem(T, (threads,))
 
     # iterate the main dimension using threads and the first block dimension
-    i = (blockIdx().x-1) * blockDim().x + threadIdx().x
+    tid = (blockIdx().x-1) * blockDim().x + threadIdx().x
     # iterate the other dimensions using the remaining block dimensions
-    j = (blockIdx().z-1) * gridDim().y + blockIdx().y
+    bid = (blockIdx().z-1) * gridDim().y + blockIdx().y
 
-    if j > length(Rother)
+    lane_id = (threadIdx().x - 1) % 32
+    warp_id = (threadIdx().x - 1) ÷ 32
+
+    if bid > length(Rother)
         return
     end
 
     @inbounds begin
-        I = Rother[j]
+        I = Rother[bid]
         Ipre = Rpre[I[1]]
         Ipost = Rpost[I[2]]
     end
 
     # load input into shared memory (apply `op` to have the correct type)
-    @inbounds temp[thread] = if i <= length(Rdim)
-        op(neutral, input[Ipre, i, Ipost])
+    value = neutral
+    if tid <= length(Rdim)
+        @inbounds value = convert(T, input[Ipre, tid, Ipost])
     else
-        op(neutral, neutral)
+        return
     end
 
-    # build sum in place up the tree
-    offset = 1
-    d = threads>>1
-    while d > 0
-        sync_threads()
-        @inbounds if thread <= d
-            ai = offset * (2*thread-1)
-            bi = offset * (2*thread)
-            temp[bi] = op(temp[ai], temp[bi])
+    sync_threads()
+
+    # Compute sums within thread warp
+    i = 1
+    while i <= 32
+        n = shfl_up_sync(mask, value, i)
```

Comment:
This breaks on older devices that don't have `shfl`. See our mapreduce implementation which has a fallback for those devices.

Reply:
What's the minimum compute capability do we plan to support ? `SHFL` intrinsics are supported from 3.0 which seemed like a decent starting point to me. I am guessing support for any prior compute capability is only for people on old computers with integrated laptops where speed is not a priority.

Reply:
Maybe it's time to bump it to 3.0, but then we should either warn about that during `__init__`, or error out when calling this function. I'll see about that in another PR.

Reply:
To put it in perspective with the year of release:
3.0 : Kepler Architecture : 2012 release
5.0 : Maxwell Architecture: 2014 release

Have a look [here](https://en.wikipedia.org/wiki/CUDA#Version_features_and_specifications), something like 3.0 or 3.5 as the minimum version makes sense to me. I doubt there are users serious about running on 3.0 compute capability.

Reply:
There are likely people out there who are using an old K40/K50 those were very popular cards.

Reply:
K40 is 3.5 . But how many people are going to be using anything made before that ?

Reply:
OK, fair enough, I'm happy to bump the required capability. but `shfl` is only supported for certain types, `(Bool, Int32, Int64, Float32, Float64, ComplexF32, ComplexF64)`, so we need a fallback anyhow.